src/dbnode/encoding/proto/encoder.go

```diff
@@ -0,0 +1,910 @@
+	enc.lastEncoded = nil
+	enc.lastEncodedDP = ts.Datapoint{}
+	enc.unmarshaled = nil
+
+	if enc.schema != nil {
+		enc.customFields = resetCustomFields(enc.customFields, enc.schema)
+	}
+
+	enc.hasEncodedFirstSetOfCustomValues = false
+	enc.closed = false
+	enc.numEncoded = 0
+}
+
+func (enc *Encoder) resetSchema(schema *desc.MessageDescriptor) {
+	enc.schema = schema
+	enc.customFields = resetCustomFields(enc.customFields, enc.schema)
+
+	enc.lastEncoded = dynamic.NewMessage(schema)
+	enc.unmarshaled = dynamic.NewMessage(schema)
+}
+
+// Close closes the encoder.
+func (enc *Encoder) Close() {
+	if enc.closed {
+		return
+	}
+
+	enc.Reset(time.Time{}, 0)
+	enc.stream.Reset(nil)
+	enc.closed = true
+
+	if pool := enc.opts.EncoderPool(); pool != nil {
+		pool.Put(enc)
+	}
+}
+
+// Discard closes the encoder and transfers ownership of the data stream to
+// the caller.
+func (enc *Encoder) Discard() ts.Segment {
+	segment := enc.discard()
+	// Close the encoder since its no longer needed
+	enc.Close()
+	return segment
+}
+
+// DiscardReset does the same thing as Discard except it also resets the encoder
+// for reuse.
+func (enc *Encoder) DiscardReset(start time.Time, capacity int) ts.Segment {
+	segment := enc.discard()
+	enc.Reset(start, capacity)
+	return segment
+}
+
+func (enc *Encoder) discard() ts.Segment {
+	return enc.segment(false)
+}
+
+// Bytes returns the raw bytes of the underlying data stream. Does not
+// transfer ownership and is generally unsafe.
+func (enc *Encoder) Bytes() ([]byte, error) {
+	if unusableErr := enc.isUsable(); unusableErr != nil {
+		return nil, unusableErr
+	}
+
+	bytes, _ := enc.stream.Rawbytes()
+	return bytes, nil
+}
+
+func (enc *Encoder) encodeCustomValues(m *dynamic.Message) error {
+	for i, customField := range enc.customFields {
+		iVal, err := m.TryGetFieldByNumber(customField.fieldNum)
+		if err != nil {
+			return fmt.Errorf(
+				"%s error trying to get field number: %d",
+				encErrPrefix, customField.fieldNum)
+		}
+
+		customEncoded := true
+		switch {
+		case isCustomFloatEncodedField(customField.fieldType):
+			if err := enc.encodeTSZValue(i, iVal); err != nil {
+				return err
+			}
+		case isCustomIntEncodedField(customField.fieldType):
+			if err := enc.encodeIntValue(i, iVal); err != nil {
+				return err
+			}
+		case customField.fieldType == cBytes:
+			if err := enc.encodeBytesValue(i, iVal); err != nil {
+				return err
+			}
+		default:
+			customEncoded = false
+		}
+
+		if customEncoded {
+			// Remove the field from the message so we don't include it
+			// in the proto marshal.
+			m.ClearFieldByNumber(customField.fieldNum)
+		}
+	}
+	enc.hasEncodedFirstSetOfCustomValues = true
+
+	return nil
+}
+
+func (enc *Encoder) encodeTSZValue(i int, iVal interface{}) error {
+	var (
+		val         float64
+		customField = enc.customFields[i]
+	)
+	switch typedVal := iVal.(type) {
+	case float64:
+		val = typedVal
+	case float32:
+		val = float64(typedVal)
```

Comment:
I assume casting back and forth between float32 and float64 will give you the same bits?

Micro optimization here, but we may want to think about changing the floatXORState.WriteLoad to take in bits instead of a float64 such that for float32, you can XOR with half the amount of bits. I assume that will generally lead to smaller XORs and save some memory and CPU. This may not make a huge impact if you have byte slice fields, but I can imagine use cases where you only have a few floats and enums for example. This could potentially impact the compression algo there quite a bit and would save the cast cost since it's not straightforward to convert between the two.

Reply:
I can't imagine that XORing float64s is any slower than XORING float32s if you're on a 64 bit architecture since they'll each fit in a register.

I also did some benchmarks on the type cast and I think its basically free (not even a nanosecond of difference in per operation cost when you add a type-cast from float32 to float64)

Reply:
Hmm yeah looks like the cast between the two is almost nothing - I thought it was going to be more since float32 has an 8 bit exponent and 23 bit fraction whereas float64 has a 11 bit exponent and 52 bit fraction so it's not like you can simple chop off some bits, the underlying bits are quite different. 

Might be worth a benchmark to see if there's any difference in compression ratio between XORing the two when the input value set is all float32 and changing. Would be good evidence as to the design choice here either way.